In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV, cross_val_score, KFold  # Import KFold
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
import statsmodels.api as sm 

In [2]:
import statsmodels.api as sm

# PART A

In [3]:
df= pd.read_csv('/Users/sunnyli/Desktop/dataset_lm (1).csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

In [5]:
df.head(5)

Dependent Var  Explanatory Var #1  Explanatory Var #2  Explanatory Var #3   
0      56.293458           13.698667           50.639873                   0  \
1      58.473431            2.714725           65.845845                   1   
2      94.195330           11.618072           65.072497                   0   
3      29.074583            0.818623           45.408996                   1   
4      86.035569            9.077544           73.548021                   0   

   Explanatory Var #4  Explanatory Var #5  Explanatory Var #6   
0          -18.568035           45.121911           11.412501  \
1          -25.105932           47.190213           10.080280   
2           -7.897464           52.163036           11.057301   
3          -18.316132           54.356714            5.029029   
4          -19.204165           47.186807           12.128134   

   Explanatory Var #7  Explanatory Var #8  Explanatory Var #9   
0           56.410757                   2          -12.281132  \
1           65.383107                   3          -36.763585   
2           82.812717                   0          -15.733547   
3           48.812471                   1          -12.825591   
4           62.520911                   2          -13.804860   

   Explanatory Var #10  Explanatory Var #11  Explanatory Var #12   
0            38.996909            -3.010548            49.195073  \
1            51.654939             4.991111            45.591729   
2            48.913837            -2.457696            56.608806   
3            45.851732            14.974177            47.362594   
4            47.765904             9.593982            53.700562   

   Explanatory Var #13  Explanatory Var #14  Explanatory Var #15  
0                    0           -21.153143            46.919314  
1                    0            -6.474403            53.383508  
2                    0           -27.903299            48.515026  
3                    1           -10.064411            55.266254  
4                    0           -17.546302            48.150543

In [6]:
mu, sigma = 0, 5 # mean and standard deviation of normal distribution for the error term
x = np.random.uniform(40,80,100)
epsilon = np.random.normal(mu,sigma,100)
y = 3 + 4*x + epsilon

In [7]:
y = df['Dependent Var']
X = df[['Explanatory Var #1', 'Explanatory Var #2', 'Explanatory Var #3', 'Explanatory Var #4', 'Explanatory Var #5',
          'Explanatory Var #6', 'Explanatory Var #7', 'Explanatory Var #8', 'Explanatory Var #9', 'Explanatory Var #10',
          'Explanatory Var #11', 'Explanatory Var #12', 'Explanatory Var #13', 'Explanatory Var #14', 'Explanatory Var #15']]


In [8]:
model_reg = sm.OLS(y,X).fit()
model_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Sun, 29 Oct 2023   Prob (F-statistic):                        0.00
Time:                        20:23:03   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Explanatory Var #1      1.3078      0.013    102.159      0.000       1.283       1.333
Explanatory Var #2      1.7667      0.009    203.101      0.000       1.750       1.784
Explanatory Var #3      6.6083      0.177     37.232      0.000       6.259       6.957
Explanatory Var #4      2.0725      0.011    189.896      0.000       2.051       2.094
Explanatory Var #5     -0.7776      0.011    -68.372      0.000      -0.800      -0.755
Explanatory Var #6      0.0124      0.008      1.595      0.111      -0.003       0.028
Explanatory Var #7      0.0286      0.006      4.890      0.000       0.017       0.040
Explanatory Var #8      0.3510      0.080      4.398      0.000       0.194       0.508
Explanatory Var #9     -0.0293      0.010     -2.838      0.005      -0.050      -0.009
Explanatory Var #10     0.1292      0.011     11.409      0.000       0.107       0.152
Explanatory Var #11     0.0145      0.013      1.152      0.250      -0.010       0.039
Explanatory Var #12     0.0752      0.009      8.698      0.000       0.058       0.092
Explanatory Var #13     0.7414      0.177      4.198      0.000       0.394       1.089
Explanatory Var #14    -0.0218      0.011     -1.974      0.049      -0.043   -8.61e-05
Explanatory Var #15     0.1212      0.012     10.324      0.000       0.098       0.144
==============================================================================
Omnibus:                        0.087   Durbin-Watson:                   1.875
Prob(Omnibus):                  0.957   Jarque-Bera (JB):                0.188
Skew:                          -0.004   Prob(JB):                        0.910
Kurtosis:                       2.897   Cond. No.                         287.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
x_updated = sm.add_constant(X)
model_updated = sm.OLS(y,x_updated).fit()
model_updated.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 7.593e+30
Date:                Sun, 29 Oct 2023   Prob (F-statistic):               0.00
Time:                        20:23:03   Log-Likelihood:                 12339.
No. Observations:                 422   AIC:                        -2.465e+04
Df Residuals:                     406   BIC:                        -2.458e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   4.32e-14    7.4e+14      0.000      32.000      32.000
Explanatory Var #1      1.3000   3.49e-16   3.72e+15      0.000       1.300       1.300
Explanatory Var #2      1.7000   2.54e-16    6.7e+15      0.000       1.700       1.700
Explanatory Var #3      6.2000   4.87e-15   1.27e+15      0.000       6.200       6.200
Explanatory Var #4      2.1000      3e-16      7e+15      0.000       2.100       2.100
Explanatory Var #5     -0.9000   3.51e-16  -2.56e+15      0.000      -0.900      -0.900
Explanatory Var #6  -3.842e-16   2.12e-16     -1.812      0.071   -8.01e-16    3.27e-17
Explanatory Var #7  -9.626e-16   1.64e-16     -5.876      0.000   -1.28e-15   -6.41e-16
Explanatory Var #8  -2.789e-15   2.23e-15     -1.253      0.211   -7.17e-15    1.59e-15
Explanatory Var #9   7.033e-16   2.85e-16      2.471      0.014    1.44e-16    1.26e-15
Explanatory Var #10 -5.371e-16   3.55e-16     -1.514      0.131   -1.23e-15     1.6e-16
Explanatory Var #11 -1.433e-15   3.43e-16     -4.178      0.000   -2.11e-15   -7.59e-16
Explanatory Var #12  1.064e-15   2.57e-16      4.146      0.000     5.6e-16    1.57e-15
Explanatory Var #13 -1.977e-14   4.92e-15     -4.020      0.000   -2.94e-14   -1.01e-14
Explanatory Var #14 -5.991e-16   3.02e-16     -1.985      0.048   -1.19e-15   -5.64e-18
Explanatory Var #15 -3.694e-15    3.6e-16    -10.274      0.000    -4.4e-15   -2.99e-15
==============================================================================
Omnibus:                        1.606   Durbin-Watson:                   1.878
Prob(Omnibus):                  0.448   Jarque-Bera (JB):                1.555
Skew:                           0.149   Prob(JB):                        0.460
Kurtosis:                       2.990   Cond. No.                     2.55e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [10]:
epsilon[0] = np.random.normal(mu,sigma,1)
for i in range(0,99):
    epsilon[i+1]=0.4*epsilon[i]+0.6*np.random.normal(mu,sigma,1)

In [11]:
y = 3 + 4*x + epsilon

In [12]:
x_updated = sm.add_constant(x)
model_OLS = sm.OLS(y,x_updated).fit()
model_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.996
Model:                            OLS   Adj. R-squared:                  0.996
Method:                 Least Squares   F-statistic:                 2.563e+04
Date:                Sun, 29 Oct 2023   Prob (F-statistic):          2.34e-120
Time:                        20:23:03   Log-Likelihood:                -248.42
No. Observations:                 100   AIC:                             500.8
Df Residuals:                      98   BIC:                             506.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.8109      1.527      0.531      0.597      -2.219       3.841
x1             4.0193      0.025    160.085      0.000       3.969       4.069
==============================================================================
Omnibus:                        1.460   Durbin-Watson:                   1.363
Prob(Omnibus):                  0.482   Jarque-Bera (JB):                0.946
Skew:                           0.058   Prob(JB):                        0.623
Kurtosis:                       3.462   Cond. No.                         317.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# PART B

In [13]:
errors = model_OLS.resid

error_std_dev = np.std(errors)

# Calculate Autocorrelation for the first three lags manually
autocorrelation_lag1 = np.corrcoef(errors[:-1], errors[1:])[0, 1]
autocorrelation_lag2 = np.corrcoef(errors[:-2], errors[2:])[0, 1]
autocorrelation_lag3 = np.corrcoef(errors[:-3], errors[3:])[0, 1]

print(f"Standard Deviation of Errors: {error_std_dev}")
print(f"Autocorrelation Lag 1: {autocorrelation_lag1}")
print(f"Autocorrelation Lag 2: {autocorrelation_lag2}")
print(f"Autocorrelation Lag 3: {autocorrelation_lag3}")

Standard Deviation of Errors: 2.9016437222086227
Autocorrelation Lag 1: 0.3090552793959339
Autocorrelation Lag 2: -0.06901228362507945
Autocorrelation Lag 3: -0.07463199425300289


In [14]:
y_pred = model_OLS.predict(x_updated)  # Predicted values from the OLS model
error_values = y - y_pred  # Calculate the residuals

# Compute standard deviation of errors
std_deviation = error_values.std()

# Create a 1D array of length 422 with a constant value (standard deviation)
sigma = np.full(len(y), std_deviation)

# Fit a GLS model using the OLS results and the calculated error values
model_gls = sm.GLS(y, x_updated, sigma=sigma)
results_gls = model_gls.fit()

# Display the summary table for the GLS model
print(results_gls.summary())

                            GLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.996
Model:                            GLS   Adj. R-squared:                  0.996
Method:                 Least Squares   F-statistic:                 2.563e+04
Date:                Sun, 29 Oct 2023   Prob (F-statistic):          2.34e-120
Time:                        20:23:03   Log-Likelihood:                -248.42
No. Observations:                 100   AIC:                             500.8
Df Residuals:                      98   BIC:                             506.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.8109      1.527      0.531      0.5

# PART C

In [17]:
from sklearn.linear_model import Lasso

In [23]:
X = df.drop('Dependent Var', axis=1)  # Explanatory variables
y = df['Dependent Var']  # Dependent variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

X_train.shape, X_test.shape

((211, 15), (211, 15))

The dataset has been split into training and test sets.
 Both sets contain 211 samples each.

In [24]:
# Running the Lasso regression model with alpha=1
lasso_model = Lasso(alpha=1)
lasso_model.fit(X_train, y_train)

# Extracting the coefficients
coefficients = lasso_model.coef_

coefficients

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

Some of the coefficients are exactly zero, which is a characteristic of Lasso regression. 

This indicates that Lasso has performed feature selection by assigning 0 weights to certain explanatory variables, effectively excluding them from the model.

In [26]:
# Predicting using the Lasso model on the test set
y_pred = lasso_model.predict(X_test)

# Calculating MAPE
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
mape

4.432190198291582

In [27]:
# Define a range of alpha values for grid search
alpha_values = np.linspace(0.01, 10, 1000)

# Lists to store results
alpha_results = []
mape_results = []

# Grid search over alpha values
for alpha in alpha_values:
    lasso_model = Lasso(alpha=alpha)
    lasso_model.fit(X_train, y_train)
    y_pred = lasso_model.predict(X_test)
    mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
    alpha_results.append(alpha)
    mape_results.append(mape)

# Find alpha that gives the minimum MAPE
min_mape_idx = np.argmin(mape_results)
optimal_alpha = alpha_results[min_mape_idx]
optimal_mape = mape_results[min_mape_idx]

optimal_alpha, optimal_mape


(0.01, 0.04429881163346635)

The approximate value for α that minimizes the Mean Absolute Percentage Error (MAPE) on the test set is
α=0.01. The corresponding MAPE for this α value is approximately 0.0443%.

The approximate value for α that minimizes the Mean Absolute Percentage Error (MAPE) on the test set is α=0.01.

# PART D

In [28]:
data = {
    "Month": list(range(1, 26)),
    "Demand": [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144]
}

df = pd.DataFrame(data)


In [29]:
advance_demand = [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]
df["Advance_demand"] = advance_demand

In [30]:
train_data = df.iloc[:20]
test_data = df.iloc[20:]

In [31]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error

# Features and target variable
X_train = train_data[["Month", "Advance_demand"]]
y_train = train_data["Demand"]
X_test = test_data[["Month", "Advance_demand"]]
y_test = test_data["Demand"]

# Create and fit the model
alpha = 1.0  # You can adjust alpha based on your model performance
model = Ridge(alpha=alpha)
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Calculate Mean Absolute Percentage Error (MAPE)
mape = mean_absolute_error(y_test, y_pred) / y_test.mean() * 100

print(f"MAPE: {mape:.2f}%")

MAPE: 15.31%


In [32]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import cross_val_score

# Features and target variable
X_train = train_data[["Month", "Advance_demand"]]
y_train = train_data["Demand"]

# Define a range of alpha values to test
alpha_values = [0.01, 0.1, 1.0, 10.0]

# Initialize variables to store the best alpha and the corresponding MAPE
best_alpha = None
best_mape = float('inf')

# Iterate through alpha values
for alpha in alpha_values:
    # Create and fit the model with the current alpha
    model = Ridge(alpha=alpha)
    
    # Perform cross-validation to estimate model performance
    mape_scores = -cross_val_score(model, X_train, y_train, scoring='neg_mean_absolute_error', cv=5)
    
    # Calculate the average MAPE score
    average_mape = mape_scores.mean()
    
    # Check if this alpha provides a better model
    if average_mape < best_mape:
        best_alpha = alpha
        best_mape = average_mape

# Fit the final model with the best alpha
final_model = Ridge(alpha=best_alpha)
final_model.fit(X_train, y_train)

# Make predictions
X_test = test_data[["Month", "Advance_demand"]]
y_test = test_data["Demand"]
y_pred = final_model.predict(X_test)

# Calculate MAPE on the test set
mape = mean_absolute_error(y_test, y_pred) / y_test.mean() * 100

print(f"Best Alpha: {best_alpha}")
print(f"MAPE on Test Data: {mape:.2f}%")

Best Alpha: 10.0
MAPE on Test Data: 15.33%


In [33]:
from sklearn.linear_model import Ridge

best_alpha = 10.0  
final_model = Ridge(alpha=best_alpha)
final_model.fit(X_train, y_train) 

Ridge(alpha=10.0)

In [34]:
y_pred = final_model.predict(X_test) 
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
print(f"MAPE on Test Data: {mape:.2f}%")

MAPE on Test Data: 15.81%
